sklearn/inspection/tests/test_permutation_importance.py

```diff
@@ -150,3 +155,65 @@ def test_permutation_importance_linear_regresssion():
                                      scoring='neg_mean_squared_error')
     assert_allclose(expected_importances, results.importances_mean,
                     rtol=1e-1, atol=1e-6)
+
+
+def test_permutation_importance_equivalence_sequential_parallel():
+    # regression test to make sure that sequential and parallel calls will
+    # output the same results.
+    X, y = make_regression(n_samples=500, n_features=10, random_state=0)
+    lr = LinearRegression().fit(X, y)
+
+    importance_sequential = permutation_importance(
+        lr, X, y, n_repeats=5, random_state=0, n_jobs=1
+    )
+
+    # First check that the problem is structured enough and that the model is
+    # complex enough to not yield trivial, constant importances:
+    imp_min = importance_sequential['importances'].min()
+    imp_max = importance_sequential['importances'].max()
+    assert imp_max - imp_min > 0.3
+
+    # The actually check that parallelism does not impact the results
+    # either with shared memory (threading) or without isolated memory
+    # via process-based parallelism using the default backend
+    # ('loky' or 'multiprocessing') depending on the joblib version:
+
+    # process-based parallelism (by default):
+    importance_processes = permutation_importance(
+        lr, X, y, n_repeats=5, random_state=0, n_jobs=2)
+    assert_allclose(
+        importance_processes['importances'],
+        importance_sequential['importances']
+    )
+
+    # thread-based parallelism:
+    with parallel_backend("threading"):
+        importance_threading = permutation_importance(
+            lr, X, y, n_repeats=5, random_state=0, n_jobs=2
+        )
+    assert_allclose(
+        importance_threading['importances'],
+        importance_sequential['importances']
+    )
```

Comment:
This test should probably be extended to also check the same property when we pass `X_df = pd.DataFrame(X)` to ensure that the shuffling of the column view works as expected. I am not 100% sure about the semantics of `.values` on a dataframe column.

Reply:
I think that we should have mixed type for different column and even mixed type in a single column. My intuition is that each column should be in the same memory block in pandas so we should still get a view

Reply:
I am not sure what you mean by "even mixed type in a single column", `pandas.Series` have a single dtype.

Anyway I would expect the `.values` of a Series to always return a view although I am not sure what it does with non-numpy-native dtypes...

The documentation is not very explicit:

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.values.html#pandas.Series.values

The newly introduced `.array` attribute should always return a view:

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.values.html#pandas.Series.array

However this is not supported in pandas < 0.24.

In case we have a column-transformers with custom transformers that extract features from non-numpy-native dtyped columns, the current code might not do what we want. I think we should write a new test for this.

Reply:
Maybe @jorisvandenbossche could shed some light? ;)

Reply:
> In case we have a column-transformers with custom transformers that extract features from non-numpy-native dtyped columns, the current code might not do what we want. I think we should write a new test for this.

Actually a categorical dtyped column + `OrdinalEncoder` or `OneHotEncoding` would be enough to test this case without introducing new custom feature engineering code.

Reply:
It seems to work as a view:

```python
>>> import numpy as np                                                                 
>>> import pandas as pd                                                                
>>> col = pd.Series(pd.Categorical(['a', 'a', 'b', 'b', 'c', 'c']))                    
>>> col                                                                                
0    a
1    a
2    b
3    b
4    c
5    c
dtype: category
Categories (3, object): [a, b, c]
>>> col.values                                                                         
[a, a, b, b, c, c]
Categories (3, object): [a, b, c]
>>> np.random.shuffle(col.values)                                                      
>>> col                                                                                
0    b
1    a
2    c
3    c
4    a
5    b
dtype: category
Categories (3, object): [a, b, c]
```

Reply:
for the safest way is to do:

```python
    # column_data = _safe_indexing(X_permuted, col_idx, axis=1)
    # column_data = getattr(column_data, "values", column_data)
    scores = np.zeros(n_repeats)
    for n_round in range(n_repeats):
        shuffling_idx = random_state.choice(
            np.arange(X.shape[0]), size=X.shape[0], replace=False
        )
        if hasattr(X_permuted, "iloc"):
            X_permuted.iloc[:, col_idx] = X_permuted.iloc[shuffling_idx, col_idx]
        else:
            X_permuted[:, col_idx] = X_permuted[shuffling_idx, col_idx]
        # random_state.shuffle(column_data)
        feature_score = scorer(estimator, X_permuted, y)
        scores[n_round] = feature_score
```